etna/transforms/math/lags.py

```diff
@@ -96,3 +100,205 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
+
+            if horizon < 1:
+                raise ValueError(f"{self.__class__.__name__} works only with positive horizon values, {horizon} given")
+
+            self.horizon = horizon
+            self._auto = True
+
+    def _save_exog_last_date(self, df_exog: pd.DataFrame):
+        """Save last available date of each exogenous variable."""
+        exog_names = set(df_exog.columns.get_level_values("feature"))
+
+        self._exog_last_date = dict()
+        for name in exog_names:
+            feature = df_exog.loc[:, pd.IndexSlice[:, name]]
+
+            na_mask = pd.isna(feature).any(axis=1)
+            last_date = feature.index[~na_mask].max()
+
+            self._exog_last_date[name] = last_date
+
+    def fit(self, ts: TSDataset) -> "ExogShiftTransform":
+        """Fit the transform.
+
+        Parameters
+        ----------
+        ts:
+            Dataset to fit the transform on.
+
+        Returns
+        -------
+        :
+            The fitted transform instance.
+        """
+        self._freq = ts.freq
+        df_exog = ts.df_exog
+
+        if df_exog is not None:
+            self._save_exog_last_date(df_exog=df_exog)
+
+        else:
+            self._exog_last_date = {}
+
+        super().fit(ts=ts)
+
+        return self
+
+    def _fit(self, df: pd.DataFrame) -> "ExogShiftTransform":
+        """Estimate shifts for exogenous variables.
+
+        Parameters
+        ----------
+        df:
+            dataframe with data.
+
+        Returns
+        -------
+        :
+            Fitted `ExogShiftTransform` instance.
+        """
+        feature_names = self._get_feature_names(df=df)
+
+        self._exog_shifts = dict()
+        self._created_regressors = []
+
+        for feature_name in feature_names:
+            shift = self._estimate_shift(df=df, feature_name=feature_name)
+            self._exog_shifts[feature_name] = shift
+
+            if shift > 0:
+                self._created_regressors.append(f"{feature_name}_shift_{shift}")
+
+        return self
+
+    def _get_feature_names(self, df: pd.DataFrame) -> List[str]:
+        """Return the names of exogenous variables."""
+        if self._exog_last_date is not None and len(self._exog_last_date) > 0:
+            feature_names = list(self._exog_last_date.keys())
+
+        else:
+            feature_names = []
+
+        df_columns = df.columns.get_level_values("feature")
+        for name in feature_names:
+            if name not in df_columns:
+                raise ValueError(f"Feature `{name}` is expected to be in the dataframe!")
+
+        return feature_names
+
+    def _estimate_shift(self, df: pd.DataFrame, feature_name: str) -> int:
+        """Estimate shift value for exogenous variable."""
+        if not self._auto:
+            return self.lag  # type: ignore
+
+        if self._exog_last_date is None or self._freq is None:
+            raise ValueError("Call `fit()` method before estimating exog shifts!")
```

Comment:
Could we really reach this `if` clause? The error message looks different compared to error message in `def _transform`.

Reply:
We usually don't expect that private methods are called directly. Ok, it doesn't look very important.